File: ui/base-tool/src/main/java/org/cru/godtools/base/tool/model/view/TextViewUtils.java
package org.cru.godtools.base.tool.model.view;

import android.content.Context;
import android.graphics.Typeface;
import android.view.Gravity;
import android.widget.TextView;

import org.cru.godtools.xml.model.Text;

import androidx.annotation.ColorInt;
import androidx.annotation.DimenRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import static android.util.TypedValue.COMPLEX_UNIT_PX;

public final class TextViewUtils {
    public static void bind(@Nullable final Text text, @Nullable final TextView view) {
        bind(text, view, null, null);
    }

    public static void bind(@Nullable final Text text, @Nullable final TextView view,
                            @Nullable @DimenRes Integer textSize, @Nullable @ColorInt Integer defaultTextColor) {
        if (view != null) {
            // set default values if null
            if (textSize == null) {
                textSize = Text.textSize(text);
            }
            if (defaultTextColor == null) {
                defaultTextColor = Text.defaultTextColor(text);
            }

            view.setText(Text.getText(text));
            view.setTypeface(getTypeface(text, view.getContext()));

            final float size = view.getContext().getResources().getDimension(textSize);
            view.setTextSize(COMPLEX_UNIT_PX, (float) (size * Text.getTextScale(text)));

            if (text != null) {
                view.setTextColor(text.getTextColor(defaultTextColor));
            } else {
                view.setTextColor(defaultTextColor);
            }

            // set the alignment for the text
            view.setGravity((view.getGravity() & Gravity.VERTICAL_GRAVITY_MASK) | Text.getTextAlign(text).mGravity);
        }
    }

    @Nullable
    private static Typeface getTypeface(@Nullable final Text text, @NonNull final Context context) {
        return text != null ? ManifestViewUtils.getTypeface(text.getManifest(), context) : null;
    }
}

File: ui/base-tool/src/main/java/org/cru/godtools/base/tool/model/view/TextViewUtils.kt
package org.cru.godtools.base.tool.model.view

import android.content.Context
import android.graphics.Paint.UNDERLINE_TEXT_FLAG
import android.graphics.Typeface.BOLD
import android.graphics.Typeface.ITALIC
import android.graphics.Typeface.NORMAL
import android.util.TypedValue
import android.view.Gravity
import android.widget.TextView
import androidx.annotation.ColorInt
import org.cru.godtools.xml.model.Text
import org.cru.godtools.xml.model.defaultTextColor
import org.cru.godtools.xml.model.textAlign
import org.cru.godtools.xml.model.textScale
import org.cru.godtools.xml.model.textSize
import splitties.bitflags.minusFlag
import splitties.bitflags.withFlag

internal fun Text?.bindTo(view: TextView, textSize: Float? = null, @ColorInt defaultTextColor: Int? = null) {
    view.text = this?.text
    view.setTypeface(this?.getTypeface(view.context), this?.typefaceStyle ?: NORMAL)
    view.paintFlags = view.paintFlags.let {
        if (Text.Style.UNDERLINE in this?.textStyles.orEmpty()) it.withFlag(UNDERLINE_TEXT_FLAG)
        else it.minusFlag(UNDERLINE_TEXT_FLAG)
    }
    val size = textScale * (textSize ?: view.context.resources.getDimension(this.textSize))
    view.setTextSize(TypedValue.COMPLEX_UNIT_PX, size.toFloat())

    val defColor = defaultTextColor ?: this.defaultTextColor
    if (this != null) {
        view.setTextColor(getTextColor(defColor))
    } else {
        view.setTextColor(defColor)
    }

    // set the alignment for the text
    view.gravity = (view.gravity and Gravity.VERTICAL_GRAVITY_MASK) or textAlign.gravity
}

private fun Text.getTypeface(context: Context) = manifest.getTypeface(context)
private val Text.typefaceStyle get() = NORMAL
    .withFlag(if (Text.Style.BOLD in textStyles) BOLD else 0)
    .withFlag(if (Text.Style.ITALIC in textStyles) ITALIC else 0)
